app/src/main/java/com/asfoundation/wallet/ui/iab/payments/common/PurchaseBonusView.kt

```diff
@@ -0,0 +1,81 @@
+package com.asfoundation.wallet.ui.iab.payments.common
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.View
+import android.widget.FrameLayout
+import com.asf.wallet.R
+import com.asfoundation.wallet.util.CurrencyFormatUtils
+import com.asfoundation.wallet.util.WalletCurrency
+import kotlinx.android.synthetic.main.layout_purchase_bonus.view.*
+import java.math.BigDecimal
+
+/**
+ * Custom view for purchase bonus
+ *
+ * *Future note:* no_bonus_msg view was removed. Hiding the header and setting bonus description
+ * achieves the same effect.
+ */
+class PurchaseBonusView : FrameLayout {
+  private val formatter = CurrencyFormatUtils()
+
+  private var showHeader: Boolean = true
+
+  constructor(context: Context) : this(context, null)
+  constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
+  constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs,
+      defStyleAttr) {
+    inflate(context, R.layout.layout_purchase_bonus, this)
+  }
+
+  fun setPurchaseBonusHeaderValue(bonus: BigDecimal, currencySymbol: String) {
+    var scaledBonus = bonus.stripTrailingZeros()
+        .setScale(CurrencyFormatUtils.FIAT_SCALE, BigDecimal.ROUND_DOWN)
+    var newCurrencyString = currencySymbol
+    if (scaledBonus < BigDecimal("0.01")) {
+      newCurrencyString = "~$currencySymbol"
+    }
+    scaledBonus = scaledBonus.max(BigDecimal("0.01"))
+    val formattedBonus = formatter.formatCurrency(scaledBonus, WalletCurrency.FIAT)
+    val bonusMessageValue = newCurrencyString + formattedBonus
+    setPurchaseBonusHeaderValue(
+        context.getString(R.string.gamification_purchase_header_part_2, bonusMessageValue))
```

Comment:
Unfortunately yes. I think what's wrong is the logic being present in CarrierVerifyPresenter (and PurchaseBonusView is right) but I'm not sure there's a better way to do it. It is repeated in the presenter because it needs to send an error message with the parsed string (see CarrierVerifyPresenter l:124~). This parsing shouldn't be the responsability of the IabErrorFragment so it leaves this as repeated code. It could be utils but I'm not sure about that either.